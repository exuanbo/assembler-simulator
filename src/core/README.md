# Core

## Assembler

## Memory

## CPU

## Bus
