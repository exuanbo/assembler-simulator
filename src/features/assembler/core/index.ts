import { Mnemonic } from '@/common/constants'
import { invariant, unsign8 } from '@/common/utils'

import {
  AssembleEndOfMemoryError,
  DuplicateLabelError,
  JumpDistanceError,
  LabelNotExistError,
} from './exceptions'
import { type Operand, OperandType, parse, type Statement } from './parser'

export type { AssemblerErrorObject } from './exceptions'
export { AssemblerError } from './exceptions'
export type { Statement } from './parser'
export type { SourceRange } from './types'

interface LabelToAddressMap {
  [labelIdentifier: string]: number
}

const getLabelToAddressMap = (statements: Statement[]): LabelToAddressMap => {
  const labelToAddressMap: LabelToAddressMap = {}
  const statementCount = statements.length
  for (let address = 0, statementIndex = 0; statementIndex < statementCount; statementIndex++) {
    const statement = statements[statementIndex]
    const { label, instruction, operands, codes } = statement
    if (label !== null) {
      if (label.identifier in labelToAddressMap) {
        throw new DuplicateLabelError(label)
      }
      labelToAddressMap[label.identifier] = address
    }
    const firstOperand = operands[0] as Operand | undefined
    if (instruction.mnemonic === Mnemonic.ORG) {
      invariant(typeof firstOperand?.code === 'number')
      address = firstOperand.code
    } else {
      // label value has not been calculated yet
      address += codes.length + (firstOperand?.type === OperandType.Label ? 1 : 0)
      if (address > 0xff && statementIndex !== statementCount - 1) {
        throw new AssembleEndOfMemoryError(statement)
      }
    }
  }
  return labelToAddressMap
}

export interface AddressToCodeMap {
  [address: number]: number
}

export interface AddressToStatementMap {
  [address: number]: Statement
}

export type AssembleResult = [AddressToCodeMap, Partial<AddressToStatementMap>]

export const assemble = (source: string): AssembleResult => {
  const statements = parse(source)
  const labelToAddressMap = getLabelToAddressMap(statements)
  const addressToCodeMap: AddressToCodeMap = {}
  const addressToStatementMap: AddressToStatementMap = {}
  const statementCount = statements.length
  for (let address = 0, statementIndex = 0; statementIndex < statementCount; statementIndex++) {
    const statement = statements[statementIndex]
    const { instruction, operands, codes } = statement
    const firstOperand = operands[0] as Operand | undefined
    if (instruction.mnemonic === Mnemonic.ORG) {
      invariant(typeof firstOperand?.code === 'number')
      address = firstOperand.code
      continue
    }
    if (firstOperand?.type === OperandType.Label) {
      if (!(firstOperand.value in labelToAddressMap)) {
        throw new LabelNotExistError(firstOperand)
      }
      const distance = labelToAddressMap[firstOperand.value] - address
      if (distance < -128 || distance > 127) {
        throw new JumpDistanceError(firstOperand)
      }
      const unsignedDistance = unsign8(distance)
      firstOperand.code = unsignedDistance
      codes.push(unsignedDistance)
    }
    const nextAddress = address + codes.length
    codes.forEach((code, codeIndex) => {
      addressToCodeMap[address + codeIndex] = code
    })
    addressToStatementMap[address] = statement
    address = nextAddress
  }
  return [addressToCodeMap, addressToStatementMap]
}
