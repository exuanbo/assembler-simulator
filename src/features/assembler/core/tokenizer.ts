import { call, parseString, trimBrackets } from '@/common/utils'

import {
  EndOfTokenStreamError,
  NotAllowedSingleQuoteError,
  UnexpectedCharacterError,
  UnterminatedAddressError,
  UnterminatedStringError,
} from './exceptions'
import type { SourceRange } from './types'

export enum TokenType {
  Whitespace = 'Whitespace',
  Comment = 'Comment',
  Colon = 'Colon',
  Comma = 'Comma',
  Digits = 'Digits',
  Register = 'Register',
  Address = 'Address',
  String = 'String',
  Unknown = 'Unknown',
}

interface TokenRule {
  type: TokenType
  pattern: RegExp
}

// prettier-ignore
const tokenRules: readonly TokenRule[] = [
  { type: TokenType.Whitespace, pattern: /\s+/ },
  { type: TokenType.Comment,    pattern: /;.*/ },
  { type: TokenType.Colon,      pattern: /:/ },
  { type: TokenType.Comma,      pattern: /,/ },
  { type: TokenType.Digits,     pattern: /\d+\b/ },
  { type: TokenType.Register,   pattern: /[a-dA-D][lL]\b/ },
  { type: TokenType.Address,    pattern: /\[.*?\]/ },
  { type: TokenType.String,     pattern: /"(?:[^\\\r\n]|\\.)*?"/ },
  { type: TokenType.Unknown,    pattern: /\w+/ },
]

const tokenRegExpSource = tokenRules.map(({ pattern }) => `(${pattern.source})`).join('|')
const createTokenRegExp = () => new RegExp(tokenRegExpSource, 'y')

export interface Token {
  type: TokenType
  value: string
  source: string
  range: SourceRange
}

const createToken = (type: TokenType, source: string, index: number): Token => {
  const value = call((): string => {
    switch (type) {
      case TokenType.Register:
      case TokenType.Unknown:
        return source.toUpperCase()
      case TokenType.Address:
        return trimBrackets(source).trim().toUpperCase()
      case TokenType.String:
        return parseString(source)
      default:
        return source
    }
  })
  const from = index
  const to = from + source.length
  return {
    type,
    value,
    source,
    range: { from, to },
  }
}

const shouldSkip = (token: Token) =>
  token.type === TokenType.Whitespace || token.type === TokenType.Comment

interface TokenStream {
  next: () => Token | null
}

const createTokenStream = (source: string): TokenStream => {
  const regexp = createTokenRegExp()
  return {
    next: () => {
      const startIndex = regexp.lastIndex
      const match = regexp.exec(source)
      if (match) {
        for (let ruleIndex = 0; ; ruleIndex++) {
          const value = match[ruleIndex + 1]
          if (value) {
            const { type } = tokenRules[ruleIndex]
            return createToken(type, value, startIndex)
          }
        }
      }
      regexp.lastIndex = startIndex
      if (startIndex === source.length) {
        // end of source
        return null
      }
      const range: SourceRange = { from: startIndex, to: startIndex + 1 }
      const char = source[startIndex]
      switch (char) {
        case '[': {
          const [value] = source.slice(startIndex).match(/^\[\s*\w*/)!
          range.to += value.length - 1
          throw new UnterminatedAddressError(value, range)
        }
        case '"': {
          const [value] = source.slice(startIndex).match(/^".*/)!
          range.to += value.length - 1
          throw new UnterminatedStringError(value, range)
        }
        case "'":
          throw new NotAllowedSingleQuoteError(range)
      }
      throw new UnexpectedCharacterError(char, range)
    },
  }
}

export interface Tokenizer {
  hasMore: () => boolean
  peek: () => Token | null
  peekNext: () => Token | null
  advance: () => void
  consume: () => Token
  match: (type: TokenType) => boolean
}

export const createTokenizer = (source: string): Tokenizer => {
  const stream = createTokenStream(source)
  const streamNext = (): Token | null => {
    let token: Token | null = null
    do {
      token = stream.next()
    } while (token && shouldSkip(token))
    return token
  }

  let current: Token | null = streamNext()
  let next: Token | null = streamNext()

  const assertCurrent = (): Token => {
    if (!current) {
      throw new EndOfTokenStreamError()
    }
    return current
  }

  const tokenizer: Tokenizer = {
    hasMore: () => {
      return !!current
    },
    peek: () => {
      return current
    },
    peekNext: () => {
      return next
    },
    advance: () => {
      current = next
      next = streamNext()
    },
    consume: () => {
      const token = assertCurrent()
      tokenizer.advance()
      return token
    },
    match: (type) => {
      const token = assertCurrent()
      if (token.type !== type) {
        return false
      }
      tokenizer.advance()
      return true
    },
  }
  return tokenizer
}
