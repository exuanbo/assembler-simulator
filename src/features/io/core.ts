export const SKIP = -1

export const MAX_PORT = 0x0f

export enum InputPort {
  SimulatedKeyboard = 0,
  Thermostat = 3,
  Keyboard = 7,
  NumericKeypad = 8,
}

export type InputData =
  | {
      content: number
      port: InputPort
    }
  | {
      content: null
      port: null
    }

export interface InputSignals {
  data: InputData
  interrupt: boolean
  // TODO: interruptVectorAddress
}

export const initialInputSignals: InputSignals = {
  data: {
    content: null,
    port: null,
  },
  interrupt: false,
}

export enum OutputPort {
  TrafficLights = 1,
  SevenSegmentDisplay = 2,
  Heater = 3,
  SnakeInMaze = 4,
  StepperMotor = 5,
  Lift = 6,
  Keyboard = 7,
  NumericKeypad = 8,
}

type OutputFlagSignalName = 'halted' | 'closeWindows'

type OutputFlagSignals = {
  [signalName in OutputFlagSignalName]?: true
}

interface OutputData {
  content: number
  port: OutputPort
}

export interface OutputSignals extends OutputFlagSignals {
  expectedInputPort?: InputPort
  data?: OutputData
}

export interface Signals {
  input: InputSignals
  output: OutputSignals
}
