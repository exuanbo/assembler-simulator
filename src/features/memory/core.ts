import { Ascii, Mnemonic } from '@/common/constants'
import type { AddressToCodeMap, AddressToStatementMap } from '@/features/assembler/core'

export type MemoryData = number[]

const MEMORY_SIZE = 0x100

export const VDU_START_ADDRESS = 0xc0

export const initVduData = (): number[] =>
  new Array<number>(MEMORY_SIZE - VDU_START_ADDRESS).fill(Ascii.Space)

export const initData = (): MemoryData =>
  new Array<number>(VDU_START_ADDRESS).fill(0).concat(initVduData())

export const initDataFrom = (map: AddressToCodeMap): MemoryData => {
  const data = initData()
  for (const address in map) {
    data[address] = map[address]
  }
  return data
}

export const getVduDataFrom = (data: MemoryData): number[] => data.slice(VDU_START_ADDRESS)

export const getSourceFrom = (map: Partial<AddressToStatementMap>): string[] => {
  const source: string[] = []
  for (let address = 0; address < MEMORY_SIZE; address++) {
    source.push(address < VDU_START_ADDRESS ? Mnemonic.END : '')
  }
  for (const address in map) {
    const statement = map[address]
    // istanbul ignore next
    if (statement === undefined) {
      continue
    }
    const { instruction, operands } = statement
    if (instruction.mnemonic === Mnemonic.DB) {
      const operand = operands[0]
      // OperandType.Number
      if (typeof operand.code === 'number') {
        source[address] = operand.value
      } else {
        // OperandType.String
        operand.value.split('').forEach((char, charIndex) => {
          source[Number(address) + charIndex] = char
        })
      }
    } else {
      source[address] = instruction.mnemonic
      const nextAddress = Number(address) + 1
      operands.forEach((operand, operandIndex) => {
        const { value } = operand
        // Address or RegisterAddress
        const isAddressOperand = operand.type.endsWith('Address')
        source[nextAddress + operandIndex] = isAddressOperand ? `[${value}]` : value
      })
    }
  }
  return source
}
