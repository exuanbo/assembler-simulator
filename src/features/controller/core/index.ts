import { produce } from 'immer'
import { first } from 'rxjs'

import { store } from '@/app/store'
import { observe } from '@/common/observe'
import { call } from '@/common/utils'
import { assemble as assembleFrom } from '@/features/assembler/assemble'
import { selectAssembledSource } from '@/features/assembler/assemblerSlice'
import {
  __getSrInterruptFlag,
  RuntimeError,
  step as stepPure,
  type StepOutput,
  type StepResult,
} from '@/features/cpu/core'
import {
  resetCpuState,
  selectCpuRegisters,
  selectCpuStatus,
  setCpuFault,
  setCpuHalted,
  setCpuRegisters,
} from '@/features/cpu/cpuSlice'
import { lineRangesOverlap } from '@/features/editor/codemirror/text'
import {
  clearEditorMessage,
  type EditorMessage,
  MessageType,
  selectEditorBreakpoints,
  selectEditorInput,
  selectEditorMessage,
  setEditorMessage,
} from '@/features/editor/editorSlice'
import { setException } from '@/features/exception/exceptionSlice'
import { InputPort, OutputPort } from '@/features/io/core'
import {
  clearInputData,
  IoDeviceName,
  resetIoState,
  selectInputSignals,
  selectIsWaitingForInput,
  setInterrupt,
  setIoDeviceData,
  setIoDevicesInvisible,
  setVduDataFrom,
  setWaitingForInput,
  setWaitingForKeyboardInput,
} from '@/features/io/ioSlice'
import { VDU_START_ADDRESS } from '@/features/memory/core'
import { resetMemoryData, selectMemoryData, setMemoryData } from '@/features/memory/memorySlice'

import {
  selectIsRunning,
  selectIsSuspended,
  selectRuntimeConfiguration,
  setRunning,
  setSuspended,
} from '../controllerSlice'
import { selectCurrentStatementRange } from '../selectors'

interface StepOptions {
  isUserAction?: boolean
  lastStepResult?: StepResult | null
}

const sourceChangedMessage: EditorMessage = {
  type: MessageType.Warning,
  content: 'Warning: Source code has changed since last assemble.',
}

export class Controller {
  private stepIntervalId?: number

  private interruptIntervalId?: number
  private isInterruptIntervalSet = false

  private lastStep: Promise<StepResult | null> = Promise.resolve(null)

  private dispatchChangesTimeoutId: number | undefined
  private get willDispatchChanges(): boolean {
    return this.dispatchChangesTimeoutId !== undefined
  }

  private unsubscribeSetSuspended: (() => void) | undefined

  private lastBreakpointLineNumber: number | undefined

  public assemble = (): void => {
    assembleFrom(store.getState(selectEditorInput))
  }

  public runOrStop = (): void => {
    if (!this.stopIfRunning()) {
      void this.run()
    }
  }

  /**
   * @returns true if stopped from running
   */
  private stopIfRunning(): boolean {
    const isRunning = store.getState(selectIsRunning)
    if (isRunning) {
      this.stop()
    }
    return isRunning
  }

  private stop(): void {
    this.clearStepInterval()
    if (this.isInterruptIntervalSet) {
      this.clearInterruptInterval()
    }
    store.dispatch(setRunning(false))
  }

  private clearStepInterval(): void {
    window.clearInterval(this.stepIntervalId)
  }

  private clearInterruptInterval(withFlag = true): void {
    window.clearInterval(this.interruptIntervalId)
    if (withFlag) {
      this.isInterruptIntervalSet = false
    }
  }

  private async run(): Promise<void> {
    store.dispatch(setRunning(true))
    this.setStepInterval()
    const lastStepResult = await this.lastStep
    if (lastStepResult !== null) {
      const isSrInterruptFlagSet = __getSrInterruptFlag(lastStepResult.cpuRegisters.sr)
      if (isSrInterruptFlagSet) {
        this.setInterruptInterval()
      }
    }
    await this.stepAsync({ isUserAction: true, lastStepResult })
  }

  private setStepInterval(): void {
    const { clockSpeed } = store.getState(selectRuntimeConfiguration)
    this.stepIntervalId = window.setInterval(this.step, 1000 / clockSpeed)
  }

  private setInterruptInterval(withFlag = true): void {
    const { timerInterval } = store.getState(selectRuntimeConfiguration)
    this.interruptIntervalId = window.setInterval(() => {
      store.dispatch(setInterrupt(true))
    }, timerInterval)
    if (withFlag) {
      this.isInterruptIntervalSet = true
    }
  }

  public stopAndRun = (): void => {
    this.pauseMainLoop()
    this.resumeMainLoop()
    this.step()
  }

  private pauseMainLoop(): void {
    this.clearStepInterval()
    if (this.isInterruptIntervalSet) {
      this.clearInterruptInterval(/* withFlag: */ false)
    }
  }

  private resumeMainLoop(): void {
    this.setStepInterval()
    if (this.isInterruptIntervalSet) {
      this.setInterruptInterval(/* withFlag: */ false)
    }
  }

  public step = (options: StepOptions = {}): void => {
    void this.stepAsync(options)
  }

  private stepAsync = async ({ isUserAction, lastStepResult }: StepOptions): Promise<void> => {
    if (lastStepResult === undefined) {
      lastStepResult = await this.lastStep
    }
    if (isUserAction) {
      store.dispatch(setInterrupt(false))
    }
    this.stepFrom(lastStepResult)
  }

  private stepFrom = (lastStepResult: StepResult | null): void => {
    switch (true) {
      case store.getState(selectEditorInput) !== store.getState(selectAssembledSource):
        store.dispatch(setEditorMessage(sourceChangedMessage))
        break
      case store.getState(selectEditorMessage) === sourceChangedMessage:
        store.dispatch(clearEditorMessage())
        break
    }
    const { fault, halted } = store.getState(selectCpuStatus)
    if (fault !== null || halted) {
      this.stopIfRunning()
      if (fault === null && halted) {
        // trigger `EditorMessage` re-render
        store.dispatch(setCpuHalted())
      }
      return
    }
    this.lastStep = new Promise((resolve) => {
      let stepOutput: StepOutput
      try {
        stepOutput = stepPure(
          lastStepResult ?? {
            memoryData: store.getState(selectMemoryData),
            cpuRegisters: store.getState(selectCpuRegisters),
          },
          store.getState(selectInputSignals),
        )
      } catch (exception) {
        this.stopIfRunning()
        if (exception instanceof RuntimeError) {
          const runtimeErrorObject = exception.toPlainObject()
          store.dispatch(setCpuFault(runtimeErrorObject))
        } else {
          store.dispatch(setException(exception))
        }
        resolve(null)
        return
      }
      const { memoryData, cpuRegisters, signals, changes } = stepOutput
      const changeAddress = changes.memoryData?.address
      const isVduBufferChanged = changeAddress !== undefined && changeAddress >= VDU_START_ADDRESS
      const dispatchChanges = (): void => {
        this.cancelDispatchChanges()
        this.dispatchChangesTimeoutId = window.setTimeout(() => {
          store.dispatch(setCpuRegisters(cpuRegisters))
          store.dispatch(setMemoryData(memoryData))
          if (isVduBufferChanged) {
            store.dispatch(setVduDataFrom(memoryData))
          }
          this.dispatchChangesTimeoutId = undefined
        })
      }
      if (!this.willDispatchChanges || isVduBufferChanged) {
        dispatchChanges()
      }
      const isRunning = store.getState(selectIsRunning)
      if (isRunning) {
        const isSrInterruptFlagChanged = changes.cpuRegisters?.sr?.interrupt ?? false
        if (isSrInterruptFlagChanged) {
          const isSrInterruptFlagSet = __getSrInterruptFlag(cpuRegisters.sr)
          if (isSrInterruptFlagSet) {
            if (!this.isInterruptIntervalSet) {
              this.setInterruptInterval()
            }
          } else {
            if (this.isInterruptIntervalSet) {
              this.clearInterruptInterval()
            }
          }
        }
      }
      const { data: inputData } = signals.input
      const { data: outputData, expectedInputPort } = signals.output
      if (signals.input.interrupt) {
        store.dispatch(setInterrupt(false))
      }
      if (signals.output.halted === true) {
        this.stopIfRunning()
        store.dispatch(setCpuHalted())
        resolve(null)
        return
      }
      if (signals.output.closeWindows === true) {
        store.dispatch(setIoDevicesInvisible())
      }
      let willSuspend = false
      if (expectedInputPort !== undefined) {
        store.dispatch(setWaitingForInput(true))
        if (inputData.content === null) {
          willSuspend = true
          if (isRunning) {
            this.pauseMainLoop()
          }
          store.dispatch(setSuspended(true))
          switch (expectedInputPort) {
            case InputPort.SimulatedKeyboard:
              store.dispatch(setWaitingForKeyboardInput(true))
              break
          }
          this.unsubscribeSetSuspended = observe(
            store.onAction(setSuspended).pipe(first()),
            // payload must be false
            () => {
              // state cannot be changed when suspended
              if (isRunning) {
                this.resumeMainLoop()
              }
              this.step()
            },
          )
        } else {
          // wrong port
          store.dispatch(clearInputData())
        }
      } else if (store.getState(selectIsWaitingForInput)) {
        // `step` called from actionListener
        store.dispatch(setWaitingForInput(false))
        store.dispatch(clearInputData())
      }
      if (outputData?.content !== undefined) {
        const { content: outputDataContent, port: outputPort } = outputData
        const ioDeviceName = call(() => {
          switch (outputPort) {
            case OutputPort.TrafficLights:
              return IoDeviceName.TrafficLights
            case OutputPort.SevenSegmentDisplay:
              return IoDeviceName.SevenSegmentDisplay
          }
        })
        if (ioDeviceName !== undefined) {
          store.dispatch(
            setIoDeviceData({
              name: ioDeviceName,
              data: outputDataContent,
            }),
          )
        }
      }
      let hasBreakpoint = false
      const breakpoints = store.getState(selectEditorBreakpoints)
      if (breakpoints.length && isRunning && !willSuspend) {
        const nextState = produce(store.getState(), (draft) => {
          draft.cpu.registers = cpuRegisters
          draft.memory.data = memoryData
        })
        const nextStatementRange = selectCurrentStatementRange(nextState)
        if (nextStatementRange) {
          const breakpointLineLoc = breakpoints.find((lineLoc) =>
            lineRangesOverlap(lineLoc, nextStatementRange),
          )
          if (breakpointLineLoc !== undefined) {
            hasBreakpoint = true
            if (breakpointLineLoc.number !== this.lastBreakpointLineNumber) {
              if (!this.willDispatchChanges) {
                dispatchChanges()
              }
              // `isRunning` is already checked
              this.stop()
              this.lastBreakpointLineNumber = breakpointLineLoc.number
            }
          }
        }
      }
      if (!hasBreakpoint && this.lastBreakpointLineNumber !== undefined) {
        this.resetBreakpointLineNumber()
      }
      resolve({ memoryData, cpuRegisters })
    })
  }

  private cancelDispatchChanges(): void {
    if (this.dispatchChangesTimeoutId !== undefined) {
      window.clearTimeout(this.dispatchChangesTimeoutId)
      this.dispatchChangesTimeoutId = undefined
    }
  }

  private resetBreakpointLineNumber(): void {
    this.lastBreakpointLineNumber = undefined
  }

  public reset = (): void => {
    this.resetSelf()
    store.dispatch(resetCpuState())
    store.dispatch(resetMemoryData())
    store.dispatch(resetIoState())
  }

  public resetSelf = (): void => {
    this.stopIfRunning()
    this.restoreIfSuspended()
    this.cancelDispatchChanges()
    this.resetBreakpointLineNumber()
    this.resetLastStep()
  }

  private restoreIfSuspended(): void {
    if (store.getState(selectIsSuspended)) {
      this.unsubscribeSetSuspended!()
      store.dispatch(setSuspended(false))
    }
  }

  private resetLastStep(): void {
    this.lastStep = Promise.resolve(null)
  }
}
