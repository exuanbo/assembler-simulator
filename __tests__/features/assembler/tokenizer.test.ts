import { createTokenizer, type Token } from '@/features/assembler/core/tokenizer'

import { shortArraySerializer } from '../../snapshotSerializers'

expect.addSnapshotSerializer(shortArraySerializer)

const tokenize = (source: string): Token[] => {
  const tokenizer = createTokenizer(source)
  const tokens: Token[] = []
  while (tokenizer.hasMore()) {
    tokens.push(tokenizer.consume())
  }
  return tokens
}

describe('tokenizer', () => {
  it('should skip whitespace', () => {
    expect(tokenize(' \t\ndone: \t\nend')).toMatchSnapshot()
  })

  it('should skip line with comment', () => {
    expect(tokenize('; this is a comment\nend')).toMatchSnapshot()
  })

  it('should skip comment at the end of the line', () => {
    expect(tokenize('done:; this is a comment\nend; this is another comment')).toMatchSnapshot()
  })

  it('should tokenize comma', () => {
    expect(tokenize(',\n,,')).toMatchSnapshot()
  })

  it('should tokenize digits', () => {
    expect(tokenize('0 01 002')).toMatchSnapshot()
  })

  it('should tokenize register', () => {
    expect(tokenize('al Bl cL DL')).toMatchSnapshot()
  })

  it('should tokenize address', () => {
    expect(tokenize('[] [00][al] [ Bl  ]')).toMatchSnapshot()
  })

  it('should throw UnterminatedAddressError when tokenizing address if closing bracket is missing', () => {
    expect(() => {
      tokenize('[00')
    }).toThrowErrorMatchingInlineSnapshot(`"Unterminated address '[00'."`)

    expect(() => {
      tokenize('[al\n')
    }).toThrowErrorMatchingInlineSnapshot(`"Unterminated address '[al'."`)

    expect(() => {
      tokenize('[Bl ')
    }).toThrowErrorMatchingInlineSnapshot(`"Unterminated address '[Bl'."`)

    expect(() => {
      tokenize('[cL \n')
    }).toThrowErrorMatchingInlineSnapshot(`"Unterminated address '[cL'."`)

    expect(() => {
      tokenize('[ DL ;')
    }).toThrowErrorMatchingInlineSnapshot(`"Unterminated address '[ DL'."`)
  })

  it('should tokenize string', () => {
    expect(tokenize('"" "this is a string" "\\"" "\\n"')).toMatchSnapshot()
  })

  it('should remove invalid escape', () => {
    expect(tokenize('"f\\o\\o"')).toMatchSnapshot()
    expect(tokenize('"\\0\\u1"')).toMatchSnapshot()
  })

  it('should throw UnterminatedStringError when tokenizing string if closing quote is missing', () => {
    expect(() => {
      tokenize('"\\"')
    }).toThrowErrorMatchingInlineSnapshot(`"Unterminated string '\\"\\\\\\\\\\"'."`)

    expect(() => {
      tokenize('"foo\nbar"')
    }).toThrowErrorMatchingInlineSnapshot(`"Unterminated string '\\"foo'."`)
  })

  it('should throw Error when tokenizing unsupported character', () => {
    expect(() => {
      tokenize('!')
    }).toThrowErrorMatchingInlineSnapshot(`"Unexpected character '!'."`)
  })
})
