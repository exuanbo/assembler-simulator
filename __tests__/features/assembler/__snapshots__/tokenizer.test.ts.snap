// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizer should remove invalid escape 1`] = `
[
  {
    "range": {
      "from": 0,
      "to": 7,
    },
    "source": "\\"f\\\\o\\\\o\\"",
    "type": "String",
    "value": "foo",
  },
]
`;

exports[`tokenizer should remove invalid escape 2`] = `
[
  {
    "range": {
      "from": 0,
      "to": 7,
    },
    "source": "\\"\\\\0\\\\u1\\"",
    "type": "String",
    "value": "0u1",
  },
]
`;

exports[`tokenizer should skip comment at the end of the line 1`] = `
[
  {
    "range": {
      "from": 0,
      "to": 4,
    },
    "source": "done",
    "type": "Unknown",
    "value": "DONE",
  },
  {
    "range": {
      "from": 4,
      "to": 5,
    },
    "source": ":",
    "type": "Colon",
    "value": ":",
  },
  {
    "range": {
      "from": 25,
      "to": 28,
    },
    "source": "end",
    "type": "Unknown",
    "value": "END",
  },
]
`;

exports[`tokenizer should skip line with comment 1`] = `
[
  {
    "range": {
      "from": 20,
      "to": 23,
    },
    "source": "end",
    "type": "Unknown",
    "value": "END",
  },
]
`;

exports[`tokenizer should skip whitespace 1`] = `
[
  {
    "range": {
      "from": 3,
      "to": 7,
    },
    "source": "done",
    "type": "Unknown",
    "value": "DONE",
  },
  {
    "range": {
      "from": 7,
      "to": 8,
    },
    "source": ":",
    "type": "Colon",
    "value": ":",
  },
  {
    "range": {
      "from": 11,
      "to": 14,
    },
    "source": "end",
    "type": "Unknown",
    "value": "END",
  },
]
`;

exports[`tokenizer should tokenize address 1`] = `
[
  {
    "range": {
      "from": 0,
      "to": 2,
    },
    "source": "[]",
    "type": "Address",
    "value": "",
  },
  {
    "range": {
      "from": 3,
      "to": 7,
    },
    "source": "[00]",
    "type": "Address",
    "value": "00",
  },
  {
    "range": {
      "from": 7,
      "to": 11,
    },
    "source": "[al]",
    "type": "Address",
    "value": "AL",
  },
  {
    "range": {
      "from": 12,
      "to": 19,
    },
    "source": "[ Bl  ]",
    "type": "Address",
    "value": "BL",
  },
]
`;

exports[`tokenizer should tokenize comma 1`] = `
[
  {
    "range": {
      "from": 0,
      "to": 1,
    },
    "source": ",",
    "type": "Comma",
    "value": ",",
  },
  {
    "range": {
      "from": 2,
      "to": 3,
    },
    "source": ",",
    "type": "Comma",
    "value": ",",
  },
  {
    "range": {
      "from": 3,
      "to": 4,
    },
    "source": ",",
    "type": "Comma",
    "value": ",",
  },
]
`;

exports[`tokenizer should tokenize digits 1`] = `
[
  {
    "range": {
      "from": 0,
      "to": 1,
    },
    "source": "0",
    "type": "Digits",
    "value": "0",
  },
  {
    "range": {
      "from": 2,
      "to": 4,
    },
    "source": "01",
    "type": "Digits",
    "value": "01",
  },
  {
    "range": {
      "from": 5,
      "to": 8,
    },
    "source": "002",
    "type": "Digits",
    "value": "002",
  },
]
`;

exports[`tokenizer should tokenize register 1`] = `
[
  {
    "range": {
      "from": 0,
      "to": 2,
    },
    "source": "al",
    "type": "Register",
    "value": "AL",
  },
  {
    "range": {
      "from": 3,
      "to": 5,
    },
    "source": "Bl",
    "type": "Register",
    "value": "BL",
  },
  {
    "range": {
      "from": 6,
      "to": 8,
    },
    "source": "cL",
    "type": "Register",
    "value": "CL",
  },
  {
    "range": {
      "from": 9,
      "to": 11,
    },
    "source": "DL",
    "type": "Register",
    "value": "DL",
  },
]
`;

exports[`tokenizer should tokenize string 1`] = `
[
  {
    "range": {
      "from": 0,
      "to": 2,
    },
    "source": "\\"\\"",
    "type": "String",
    "value": "",
  },
  {
    "range": {
      "from": 3,
      "to": 21,
    },
    "source": "\\"this is a string\\"",
    "type": "String",
    "value": "this is a string",
  },
  {
    "range": {
      "from": 22,
      "to": 26,
    },
    "source": "\\"\\\\\\"\\"",
    "type": "String",
    "value": "\\"",
  },
  {
    "range": {
      "from": 27,
      "to": 31,
    },
    "source": "\\"\\\\n\\"",
    "type": "String",
    "value": "
",
  },
]
`;
